### PR TITLE
Support 2D pointwise scheduler with autotuning

### PR DESCRIPTION
**Pull request by rdspring1**
This PR refactors `autotune_pointwise.py` to support 2D pointwise scheduler.

* Create `autotune_utils.py` to hold common utilities such as `ScriptConfiguration`, `collect_data`, `separate_data`, `test_model_rmse`, `test_model`
* Added support for `Gelu-Bias`, `Silu-Mul`, `Bcast-Add`, `Mul` fusions
* Use `ensemble.RandomForestRegressor` instead of `tree.DecisionTreeRegressor`

Simplified script flow into 5 steps:
1. Setup Script Configuration by creating `ScriptConfiguration` and `Autotune` class.
2. Run experiments by calling `collect_data`.
3. Separate data into training and validation set using `separate_data`.
4. Train the regression model using sklearn random forest ensemble.
5. Test the regression model using `test_model_rmse` and `test_model`.

### CODE REVIEW

**Comment by rdspring1** [comment]
!build

**Comment by rdspring1** [comment]
!build

**Comment by rdspring1** [comment]
!build

**Comment by rdspring1** [comment]
!build

**Review by liqiangxl — COMMENTED** [review]
LGTM, added minor suggestions to make the script more user fridendly.

File: doc/dev/python_scheduling/autotune_pointwise.py [hunk 1]
```diff
@@ -25,287 +52,321 @@
...
+            return outer_bcast()
+        elif self.selected_fusion in [self.FUSION.SILU_MUL, self.FUSION.MUL]:
+            return full()
+        elif self.selected_fusion == FUSION.BCAST_ADD:
+            return inner_bcast()
+        else:
+            assert False
+
+    # A decorator to create a pointwise fusion given some input arguments.
+    def create_fusion_func(self, inputs):
```
[the remark this answers is not in the document]
Comment (rdspring1) [reply doc/dev/python_scheduling/autotune_pointwise.py:164]: I do in a roundabout way. Thunder bakes static shapes into its definition. I modify the definition by hand to accept symbolic sizes.

File: doc/dev/python_scheduling/autotune_pointwise.py [hunk 2]
```diff
@@ -25,287 +52,321 @@
...
+        def _named_product(**items):
+            return itertools.starmap(
+                self.PointwiseConfiguration, itertools.product(*items.values())
+            )
+
+        num_dimensions = len(input_shape)
+        warp_size = 32
+        warp_group = warp_size * 4
+        # limited to a maximum of 128 threads because of pointwise scheduler
+        max_threads_per_cta = 128
```
[the remark this answers is not in the document]
Comment (rdspring1) [reply doc/dev/python_scheduling/autotune_pointwise.py:96]: Yes, I believe so. It didn't work when I tried larger cta sizes.

File: doc/dev/python_scheduling/autotune_pointwise.py [hunk 3]
```diff
@@ -25,287 +52,321 @@
...
+                        2**log_bdimy for log_bdimy in range(1, log2_max_bdimy + 1)
+                    ]
+
+                    bdim_shapes = [
+                        (max(warp_size, num_threads // bdimy), bdimy)
+                        for bdimy in bdimy_configs
+                    ]
+                    # total_unroll_factor is between [1, 9] given that outer and
+                    # inner unroll factors are between [1, 3].
+                    outer_unroll_range = range(1, 4)
```
[the remark this answers is not in the document]
Comment (rdspring1) [reply doc/dev/python_scheduling/autotune_pointwise.py:122]: You can't just change the `4` to the max factor you desire when you run the script.

**Comment by rdspring1** [comment]
!test

File: doc/dev/python_scheduling/autotune_pointwise.py [hunk 4]
```diff
@@ -25,287 +52,321 @@
...
+        return fd
+
+
+# Run sequence of steps to collect data, train and test model
+def main():
+    # ====================== Setup Script Configuration  =======================
+    script_config = ScriptConfiguration(
+        num_dimensions=2,
+        outer_shapes=[16384],
+        inner_shapes=[128, 1024, 4096, 16384],
```
[the remark this answers is not in the document]
Comment (rdspring1) [reply doc/dev/python_scheduling/autotune_pointwise.py:337]: > what's the meaning of these inner_shapes=[128, 1024, 4096, 16384]?

During training, the cartesian product of `outer_shapes` and `inner_shapes` is used to define the shape of the input tensor arguments.

> Can we move all these const numbers to the beginning of this script (or another appropriate location) 

This is the beginning of the script. i.e., the main function

> add comments about them, which controls what and how should the user set them.

Comments for the arguments were in the `dataclass` definition. See https://github.com/NVIDIA/Fuser/pull/3407/files#diff-d274cdcdd075f9e662b5efc801d18fd6949110cd76fb4ba8594d0440452f3297R16-R39

**Comment by rdspring1** [comment]
!build

**Comment by rdspring1** [comment]
!build

**Comment by rdspring1** [comment]
!build

**Review by liqiangxl — APPROVED** [review]
